cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/TaskWorkerHttpHandlerInternal.java

```diff
@@ -169,31 +166,49 @@ private void enablePeriodicRestart(CConfiguration cConf,
         Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION_SECOND, 0);
     int lowerBound = (int) (duration - duration * DURATION_FRACTION);
     int upperBound = (int) (duration + duration * DURATION_FRACTION);
-    if (lowerBound > 0) {
-      int waitTime =
-          (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
-      Executors.newSingleThreadScheduledExecutor(
-              Threads.createDaemonThreadFactory("task-worker-restart"))
-          .scheduleWithFixedDelay(
-              () -> {
-                if (mustRestart.get()) {
-                  // We force pod restart as the ongoing request has not finished since last
-                  // periodic restart check.
-                  stopper.accept("");
-                  return;
-                }
-                // we restart once ongoing request (which has set runningRequestCount to 1)
-                // finishes.
-                mustRestart.set(true);
-                if (runningRequestCount.compareAndSet(0, 1)) {
-                  // there is no ongoing request. pod gets restarted.
-                  stopper.accept("");
-                }
-              },
-              waitTime,
-              waitTime,
-              TimeUnit.SECONDS);
+
+    if (lowerBound <= 0) {
+      return;
+    }
+    int waitTime = (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
+
+    int taskDeadlineSeconds = cConf.getInt(
+        TaskWorker.TASK_EXECUTION_DEADLINE_SECOND,
+        0);
+
+    if (taskDeadlineSeconds < 0) {
+      LOG.info(
+          "Task deadline is {}, using {} value {} as the deadline instead.",
+          taskDeadlineSeconds,
+          Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION_SECOND, duration);
+      taskDeadlineSeconds = duration;
     }
+    int finalTaskDeadlineSeconds = taskDeadlineSeconds;
+
+    ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(
+    Threads.createDaemonThreadFactory("task-worker-restart"));
+
+    executorService.scheduleWithFixedDelay(() -> {
+      // we restart once all ongoing requests finish, i.e. runningRequestCount is 0.
+      mustRestart.set(true);
+      LOG.debug(
+          "Task worker service is about to restart in {} seconds, no new tasks will be accepted.",
+          finalTaskDeadlineSeconds);
+      if (runningRequestCount.get() == 0) {
+        stopper.accept("");
+        executorService.shutdown();
+        return;
+      }
+      try {
+        Thread.sleep(TimeUnit.SECONDS.toMillis(finalTaskDeadlineSeconds));
+      } catch (InterruptedException e) {
+        LOG.warn(
+            "Interrupted while waiting for task completion. Stopping immediately",
+            e);
+      }
+      stopper.accept("");
+      executorService.shutdown();
+    }, waitTime, waitTime, TimeUnit.SECONDS);
```

Comment:
nit: I would actually do `waitTime, finalTaskdeadlineSeconds`.  If we are interrupted in the sleep, we'll get another run faster

Reply:
Done